### ISSUE TITLE
Optimising training by lowering MSE loss weight

### ISSUE THREAD

**Issue by jkiliani** [gcp/leela-zero#1480]
Leela Chess had from the beginning of its implementation set up two parameters policy_loss_weight and value_loss_weight in training.py for the purpose of adjusting the total training loss term relative to each other and to the L2 regularisation term. Up until recently on the 192x15 network, those weights were kept constant at 1 in line with Deepmind's parameter setting. This continued until the value head strongly overfit recently, with an accompanying strength regression since Leela chess uses no gating.

At the height of the regression, learning rates were lowered which arrested the progression of the overfit value head, but did not reverse it. Then @Tilps had the idea to set value_loss_weight to 0.25 from 1, and the value head immediately began to recover along with playing strength. This happened in spite of continuing overtraining on single positions in a setup without gating, it was calculated that any single position was used ~12 times per training.

More recently, the sampling was reduced by training in intervals with training steps strictly correlating to rate of game generation, while resetting value_loss_weight to 1. This seems to also be effective in producing a good value head, but policy training has slowed down a lot since.

Recent self-play Elo progress and Tensorboard charts on Leela Chess MSE, policy loss and Reg term during the period in question:

![tensorboard](https://user-images.githubusercontent.com/33714573/40563078-c1105c9c-6063-11e8-9e7e-bbfa6d29f5d5.png)

Prior to step count 2.15M, cyclic learning rate schedule of 25k steps at 0.001 and 25k steps at 0.0005 was used. Up to step 2.5M, the learning rates were lowered to 0.0005/0.0001, at the same step count schedule. At 2.5M, value_loss_weight was reduced. 11.8M games on the self-play Elo chart correspond to the start of the Tensorboard charts (marked green on Elo chart), the change in value_loss_weight was at the precise turning point at 12.93M games.

Here is my theory on the regularisation ratio (i.e. value_loss_weight/policy_loss_weight) from the observations in Leela chess, I also posted this in Discord:

We’re training the policy head to approximate the visit count distribution and would like to have it ideally become that distribution. Based on that, overfitting the policy head seems very hard to do, since more training on it will just replace the previous policy with a better one (e.g. including moves found by Dirichlet noise), so moderately high learning rates and many steps should not do any harm regarding policy alone. This would be very different when training policy toward a one-hot vector, in that case policy overfitting would actually happen.

The value head on the other hand we train toward the binary game outcome, but we do NOT want it to actually fit that, since that would be a catastrophically overfit evaluation. Instead, we want the value head to assume a good statistic on all possible game outcomes from positions similar to the current one. To achieve that, only very low learning rates and high regularisation succeed in training the outcome toward the one in the current game, without forgetting the almost equally valuable game outcomes in similar positions from slightly older games.

The only way to strongly train policy, to get good feedback regarding Dirichlet noise, while weakly training value, is to adjust the regularisation ratio to the current progress in training. Very early in training, i.e. from a random net, a regularisation ratio even larger than 1 would probably have worked for a while, since training the policy is completely ineffective without a useful evaluation. But the more accurate board evaluations get, and the slower strength progress, the less training on value will be actually useful, or the quality of the statistic on game outcomes will not improve from it. So the optimal regularisation ratio should decrease over time, with stronger networks.

Unlike Leela Chess, Leela Zero uses gating which means that the value head cannot strongly overfit for a network that can pass gating. The flip side of this approach is that the amount of training is limited to how much the value head can tolerate, which is almost certainly much less than would work for the policy head. Using a smaller weight on MSE loss in the total loss calculation would allow either much longer training runs before value started to overfit, or to use higher learning rates with more policy gains when using the same number of training steps, which would radically improve the utilisation of the policy training data and shorten the learning curve. Unlike e.g. #1197, such a change would be fully compatible with SWA, and would also make even an always-promote strategy for Leela Zero much less risky if we decided to do so at some point, since it appears to strongly increase the tolerance of the value head for training more steps or at larger learning rates.

**Comment by gcp** [comment]
Lowering the value loss weight was very obvious because it's already proposed in the very first Alpha Go Zero paper for using supervised learning. Clearly they did not feel this was necessary to do for the reinforced learning, though, not for Go and as far as we know not for chess either. (I am not entirely clear on why - there is obviously 400 times as much policy training info as there is value info, and this ratio is much worse for Go compared to chess.)

I don't think we're currently seeing any overfitting on the value side.

>The value head on the other hand we train toward the binary game outcome, but we do NOT want it to actually fit that, since that would be a catastrophically overfit evaluation.

I'm not sure I agree here. Just see the how the ELF weights behave. We're trying to get rid of that because it's bad for handicap, but it clearly does not seem to prevent from achieving a high playing strength.

Also if this were true, scaling the training target with the game length would work, but all experimentation has shown that it does not.

I am everything but convinced by the argumentation, let alone changing that magic value.

**Comment by roy7** [comment]
This is kinda off topic but the post reminded me of something. A couple years back at UEC Cup a Go AI (Mojo?) got an award for originality, they did an approach using NNs but instead of training a win/loss value head it trained final color of each board position. So to get a value estimate of a board, you'd calculate a score from the NN prediction (I'm guessing).

Would something like that work in our Zero approach? We'd have as much NN power for final position colors as we do for policy, instead of 1/400 as much predicting only the win rate %.

**Comment by gcp** [comment]
It's been proposed in previous papers, but it means that you can no longer resign games. 

Also, the value estimate must necessarily come from calculating this internally in the network, so you might be disappointed to find it does not have much regularization power.

**Comment by roy7** [comment]
> It's been proposed in previous papers, but it means that you can no longer resign games.

Hmmm, interesting point. But one could just learn 19x19 final board colors + winrate then. When doing search the question might then be which is better for MCTS search and decision making, win rate or final score. I have no idea. But with 400x the learning power available, one would think final state prediction would be be more accurate than the win rate alone.

**Comment by gcp** [comment]
>win rate or final score. I have no idea.

Win rate is significantly stronger.

**Comment by jkiliani** [comment]
I agree the argumentation is somewhat speculative, and there is disagreement on why this works. However, that it worked to recover from a catastrophically overfit value head without gating, and that there were large policy gains during that recovery period is not disputed.

> I don't think we're currently seeing any overfitting on the value side.

I don't think so either, it wouldn't happen since we're using gating. But for a gating approach, this may have a different benefit, since using a smaller weight on MSE loss in the total loss calculation would allow either longer training runs or higher learning rates before value started to overfit.

> I'm not sure I agree here. Just see the how the ELF weights behave.

The ELF weights behave very well in situations that they know well, i.e. balanced ones. By training the weights for a long time, on low learning rates and with gating, the value head should becomes very strong for balanced play, is that not what we're actually seeing with ELF?

> Also if this were true, scaling the training target with the game length would work, but all experimentation has shown that it does not.

How would one scale the value training target with game length?

> I am everything but convinced by the argumentation, let alone changing that magic value.

This proposal isn't actually aimed at changing the end result, i.e. the final product of reinforcement learning at all. It's about efficiency in getting there. Deepmind's magic value is fine if efficiency in using training games is not an issue, but it's highly unlikely to be anywhere close to optimal. The argumentation may be flawed, but the data does show that changing the number can have very positive effects.

**Comment by gcp** [comment]
>The ELF weights behave very well in situations that they know well, i.e. balanced ones

They also behave correctly in that they will call 0% or 100%. Your argument was:

>The value head on the other hand we train toward the binary game outcome, but we do NOT want it to actually fit that.

And I said the above as a refutation of it.

>How would one scale the value training target with game length?

If white wins, score the opening at 50% and the last move as 100%, then scale linearly in between. This reduces overfitting because the network can't just give the 100% score if it recognizes the game by some characteristic patterns.

>Deepmind's magic value is fine if efficiency in using training games is not an issue, but it's highly unlikely to be anywhere close to optimal

This is true, but as mentioned in the my reply, they *were* aware of the possibility to scale the MSE loss, and they chose to not do so for the reinforced run.

**Comment by Eddh** [comment]
> Win rate is significantly stronger.

I'm curious about one thing : were tests done about winrate vs score with full NN guided tree search ? Before that it used Monte Carlo rollouts and that's where it was demonstrated that winrate is stronger right ? But the recent developments with AlphagoZero might change that. After all there is some evidence Golaxy predicts end score on top of/instead of winrate :  https://translate.google.com/translate?sl=auto&tl=en&js=y&prev=_t&hl=fr&ie=UTF-8&u=http%3A%2F%2Fwww.yikeweiqi.com%2Fbanner%2F43722%2F&edit-text=&act=url

**Comment by gcp** [comment]
I think most experiments date to the rollout area, but the underlying reason doesn't look like it would have changed. The problem is that you if you have a certain win by a small margin, and a much more unlikely win by a large margin, playing towards maximal territory chooses the second line, which is not what you want.

Predicting end score is useful from an analysis point of view, but that doesn't necessarily mean it makes the program stronger.

**Comment by gcp** [comment]
>Using a smaller weight on MSE loss in the total loss calculation would allow either much longer training runs before value started to overfit, or to use higher learning rates with more policy gains when using the same number of training steps

I don't understand. If we'd be using a higher learning rate than we are now, there would be less policy gains (because it would oscillate) and it's unclear what'd happen with the MSE gain (because it effectively has a lowered learning rate it could gain, but it'd be pulled around by weight changes from the policy loss).

It's exactly the opposite from what you say?

**Comment by jkiliani** [comment]
> And I said the above as a refutation of it.

A refutation that I do not think is correct. Let's assume a board pattern that occurs relatively rarely, but is balanced in that black and white win the games where it occurs equally often. Let's say we have that pattern occur only very few times (<10) in the current training window. Then training the value head with a high training momentum will cause the evaluation of that position to oscillate between significantly better than 0.5 and significantly worse, depending on whether the net was last trained on a position where that pattern led to a win, or a loss. SWA will reduce the impact of this of course, but in cases where the current training window only contains an incorrectly biased sample of game outcomes for such a board pattern, training will increasingly degrade the value head on this position.

Smaller learning rates increase the memory of a network, and a deviation from the standard value loss just allows training policy more before such problems occur.

**Comment by jkiliani** [comment]
Why would there be less policy gains from higher learning rates? Have you seen policy actually really overfit when not training it to one-hot vectors?

**Comment by gcp** [comment]
Yes, but you're working from the assumption that your dataset is bad to begin with. If it's good, then the training will allow you to recognize the won position from further ahead.

**Comment by gcp** [comment]
>Why would there be less policy gains from higher learning rates? Have you seen policy actually really overfit when not training it to one-hot vectors?

Because they would oscillate, not because they would overfit! We don't lower the learning rate unless the oscillation is so bad we can't make further progress.

**Comment by Eddh** [comment]
Neural networks might mitigate that problem in training, because if the game is a risky win by 30 points which might very well turn into a 50 point loss, the NN will not output 30 or -50 but something in between. While the averaging done in real time by rollouts might  be a lot more chaotic. After all the NN would kind of hedge between millions of games, a lot higher than rollouts.

**Comment by gcp** [comment]
>the NN will not output 30 or -50 but something in between. 

So it's translating the board occupancy to...a win rate?

**Comment by alreadydone** [comment]
> They also behave correctly in that they will call 0% or 100%. 

The ELF weight is poor in terms of relative (predictive) error for winrates close to 0% and 100%. See my tests in https://github.com/gcp/leela-zero/issues/1474#issuecomment-391806331.

> The problem is that you if you have a certain win by a small margin, and a much more unlikely win by a large margin, playing towards maximal territory chooses the second line, which is not what you want.

I think we could train the value head output before tanh towards the score, instead of training the output after tanh towards the binary outcome. During the tree search, also use the average of tanh(const * NN predicted score) as the Q-value; this should offer proper reward for the tree search to favor winning more than risky ways to get more points.

**Comment by Eddh** [comment]
> So it's translating the board occupancy to...a win rate?

Kind of since from what I understand "winrate" is a kind of misnomer, legacy from Monte Carlo rollouts. It seems more like a confidence estimation from the value head to me. 

Anyway It's still completely unknown if it would be better/worse/the same. But I guess it could help for avoiding value head overfitting. Which is not an issue they had in the era of Monte Carlo Rollouts too.

**Comment by gcp** [comment]
> this should offer proper reward for the tree search to favor winning more than risky ways to get more points, and we can adjust greediness easily by changing the constant. 

Yes, and you should find that the constant goes to very large values, because of the above problem.

> But I guess it could help for avoiding value head overfitting. 

I don't see how, as explained above you are now explicitly calculating something the network is necessarily already calculating internally. Where would the regularization properties come from?

**Comment by jkiliani** [comment]
I won't claim to really understand why the results I posted here came about, the causation in that explanation is somewhat weak as others in Leela Chess already pointed out. Maybe I should delete any attempt at explaining this? Let's just say that we have another one of Deepmind's famous magic numbers here, although it's hidden carefully, and that there is actual data indicating that it's suboptimal. What's more, it would be neither expensive nor risky to test it, since it would only require reducing the weight on a training run. If the reduced value is worse, then the nets from that run would simply fail gating.

But if the preference is to stick to this magic number anyway, I may as well close the issue...

**Comment by Eddh** [comment]
Might be my low layered brain can't understand the feels of the DRCNN, but I feel like it would be easier to train for a target that cover a wide range of value instead of a binary value. If I imagine myself in place of the NN I have trouble imagining how I would learn the game by being given 0 or 1 as an answer, as opposed to a more complex value that would differ for every game. In summary the training target can give information that could help training ?

**Comment by gcp** [comment]
>In summary the training target can give information that could help training ?

Yes, that might well be true. But at this point the network can count, so I'm not sure it makes a difference any more.

I would expect faster convergence for a network from scratch, though.

**Comment by Eddh** [comment]
I guess the main difference would be better handicap handling, and maybe more human-friendly style when winrate is already at 0.99 percent. Which would be useful for pros hoping to analyze yose positions to know which move is the best, even if one side has already lost for sure.

**Comment by gcp** [comment]
I think it's very useful when analyzing yes, as you can then see whether a positions is B+1 or B+2, etc. I think that's why bots like Golaxy etc have added this. (Regular Leela also has it, for the same reason)

**Comment by gcp** [comment]
>if the preference is to stick to this magic number anyway, I may as well close the issue...

I wouldn't be tempted to change this as long as there's no sign of overfitting on the value head - it doesn't seem very clear it can help anything if that's not a problem, right?

I'm not sure what other changes there are that would be partially beneficial but that could have as a side effect an overfit on the value head and that then could be corrected by this. Too early reductions of the learning rate?

**Comment by gcp** [comment]
>The ELF weight is poor in terms of relative (predictive) error for winrates close to 0% and 100%

That's not what I meant. I'm saying that if it says 0% or 100%, it's usually right, so that's valid target to train towards. You're saying that when it says 0.01% it should really be saying 0.1% in some cases. I don't think the latter can be solved or is related to over-fit control etc, it's a result of T=1 and so.

**Comment by jkiliani** [comment]
> I wouldn't be tempted to change this as long as there's no sign of overfitting on the value head - it doesn't seem very clear it can help anything if that's not a problem, right?

No, it's not very clear. However, it's at least plausible that a better value exists, but without trying to tune for this there is no chance of finding it. It looked to me that this was one potential way to increase the training efficiency per number of games, but it doesn't appear you're interested in looking into this.

Speaking of training efficiency, have you looked into @Ttl's analysis in #1416? Enabling time management especially looked really promising to me, but I didn't see you respond to his analysis...

**Comment by gcp** [comment]
>Enabling time management especially looked really promising to me, but I didn't see you respond to his analysis...

I didn't see the part about time management.

The problem of enabling it would be that it makes it harder for noised moves to be discovered along a forced variation, so we can then have *that* whole discussion again.

**Comment by jkiliani** [comment]
That's true, but shouldn't that aspect be OK as long as non-contender pruning is deactivated, and time management only truncates the search without additionally biasing it?

**Comment by gcp** [comment]
I think so yes. As long as the probability can start accumulating, it should be fine.

**Comment by walkenzoy** [comment]
I would also be interested to see components of the loss function graphed over training iterations as @jkiliani suggested.

While we are talking about loss functions, I am wondering why we are using MSE rather than cross-entropy for the value head ? This is consistent with alphago's methods, but it would seem like cross-entropy should help especially towards the extreme ends of the win probabilities ?

**Comment by gcp** [comment]
>I would also be interested to see components of the loss function graphed over training iterations as @jkiliani suggested.

http://storage.sjeng.org/zero/

>This is consistent with alphago's methods, but it would seem like cross-entropy should help especially towards the extreme ends of the win probabilities ?

Yes. For some network architectures you won't even manage to train the value head without using CE. I think practically it's a regression, not a classification task (though even that is arguable) which suggests using MSE.

The decision to use MSE there is consistent throughout all their publications, though.

This one can easily be tested by doing 2 training runs and seeing what you end up with.

**Comment by Ishinoshita** [comment]
@eddh wrote ", but I feel like it would be easier to train for a target that cover a wide range of values instead of a binary value." 

Just a thought, since the issue is closed:

As already discussed, a way to achieve a smoother value target could be to mix binary game outcome with the value accumulated by MCTS at root after 3200v. Each position of a single game would have different value target at training time (allowing higher learning rate?). This would also better account for stochasticity introduced in selfplay games: setting a 0% value target for a position evaluated at 70% wr by MCTS because a subsequent suboptimal move selected by temperature made the game flipped over does not sound appealing per se. This would amount mixing the MCTS eval with one single super-heavy rollout. 
But I also understand that using game outcome has been the key ingredient to achieve training of a value target, since the first AG paper. What pure temporal difference learning failed to achieve, the network chasing its tail and circling around, ignoring the real goal of the game. So this might be a misleading idea in the end....

**Comment by gcp** [comment]
@Ishinoshita I think the "problem" is that it requires much more generations to propagate back new discoveries about the game result to the earlier parts of the game.

**Comment by gcp** [comment]
>Just a thought, since the issue is closed

FWIW feel free to open new issues. This one could also just be reopened or filed anew if there is some *data* showing that I am wrong. (I just disagree that Leela Chess qualifies as data here!)

**Comment by sethtroisi** [comment]
I did some testing on how changing the weight affect supervised learning for minigo here: https://github.com/tensorflow/minigo/issues/350

Exactly what you expect happens:
if you weight policy more, policy learns faster but at the cost of value error.
If you weight value more, value learns slightly faster but at the cost of a lot of accuracy.
1.0 seemed like right in the sweet spot. weighting policy slightly more might speed learning up a little bit (especially in the first part of a training run) but it would be only a minor improvement.
